### ISSUE TITLE
Why is there a `restart` field in `shutdown_reply`?

### ISSUE THREAD

**Issue by gibiansky** [jupyter/jupyter_client#170]
It seems like this field should always be whatever was sent in the `shutdown_request`. Why include this field in the `shutdown_reply` at all?

**Comment by Carreau** [comment]
Hum, not sure, but I would say for alternative connected clients that haven't triggered the shutdown request, to know whether they can assume the kernel will restart and reconnect, of shutdown and stay off, in which case it can likely safely permanently disconnect.

**Comment by gibiansky** [comment]
If I understand correctly, though, the replies sent on the `shell` and `control` sockets are sent only to the single kernel that send the request (only messages sent on `iopub` are published to all frontends). If that's the case then that explanation doesn't quite work since the `shutdown_reply` is sent on `shell` or `control`...

**Comment by Carreau** [comment]
Yes, but IIRC the dealer/router has a necessary request/reply pattern,but the message are also broadcasted on PUB/SUB. For coherence it's likely the same message on both.

Example qtconsole, were slave detects the restart.

![resarting](https://cloud.githubusercontent.com/assets/335567/16980228/d5bcc684-4e19-11e6-92dc-82cb631d810d.gif)


This need to be checked and clarified in doc if the case.

**Comment by gibiansky** [comment]
That doesn't quite make sense – only `iopub` messages are broadcast on PUB/SUB, and there are no shutdown `iopub` messages as far as I can tell from the messaging spec.

At the very least the messaging spec says nothing about sending a `shutdown_reply` on `iopub`.

Definitely seems like something needs to be checked / clarified, I guess.

**Comment by Carreau** [comment]
> Definitely seems like something needs to be checked / clarified, I guess.

Yes it is broadcasted on IOPUB in ipykernel, 
and yes it is received by the qtconsole on IOpub by the QtConsole as well. 

Let's loop-in @willingc as this is documentation, @minrk, @takluyver and @fperez. It touches the protocol, so we should be careful when updating the docs if some of the implementation of ipykernel or qtconsole are mistakes.

**Comment by takluyver** [comment]
I think it makes sense for all frontends to know when the kernel is stopping or restarting. The precise implementation may well have been a hack, however.

We can also hopefully resolve this better with the kernel nanny proposal (hint, hint ;-) jupyter/enhancement-proposals#14), because the nanny can send messages after the kernel unexpectedly dies, as well as if it is deliberately shut down.

**Comment by minrk** [comment]
I'll need to read some things to refresh my memory on the details, but I seem to recall it being something of a hack for notifying other frontends when restart has been requested during early qtconsole development. I think the kernel doesn't take any action wrt whether it's a restart or not. I suspect we might be broadcasting the shutdown_reply on IOPub (in addition to the actual reply on shell), which would be a bit gross, and should probably be a dedicated `shutdown_notification` message.
